src/stream/mod.rs

```
@@ -471,3 +476,231 @@ impl Stream {
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        if buf.remaining() == 0 {
+            return Poll::Ready(Ok(()));
+        }
+
+        let fut = match self.read_fut.as_mut() {
+            Some(fut) => fut,
+            None => {
+                // read into a temporary buffer because `buf` has an unonymous lifetime, which can be
+                // shorter than the lifetime of `read_fut`.
+                let stream = self.stream.clone();
+                let mut temp_buf = vec![0; buf.remaining()];
+                self.read_fut.get_or_insert(Box::pin(async move {
+                    let res = stream.read(temp_buf.as_mut_slice()).await;
+                    match res {
+                        Ok(n) => {
+                            temp_buf.truncate(n);
+                            Ok(temp_buf)
+                        }
+                        Err(e) => Err(e),
+                    }
+                }))
+            }
+        };
+
+        loop {
+            match fut.as_mut().poll(cx) {
+                Poll::Pending => return Poll::Pending,
+                // retry immediately upon empty data or incomplete chunks
+                // since there's no way to setup a waker.
+                Poll::Ready(Err(Error::ErrTryAgain)) => {}
+                // EOF has been reached => don't touch buf and just return Ok
+                Poll::Ready(Err(Error::ErrEof)) => {
+                    self.read_fut = None;
+                    return Poll::Ready(Ok(()));
+                }
+                Poll::Ready(Err(e)) => {
+                    self.read_fut = None;
+                    return Poll::Ready(Err(e.into()));
+                }
+                Poll::Ready(Ok(read_buf)) => {
+                    let len = std::cmp::min(read_buf.len(), buf.remaining());
+                    buf.put_slice(&read_buf[..len]);
+                    self.read_fut = None;
```

If `read_buf.len()` is > to `buf.remaining()` then you're silently discarding some data in `read_buf` here

---

good catch!